src/v/serde/parquet/shredder.cc

```diff
@@ -0,0 +1,279 @@
+  traversal_levels levels,
+  absl::FunctionRef<
+    void(const indexed_schema_element&, traversal_levels, value)> cb) {
+    ss::visit(
+      std::move(val),
+      [&element, levels, cb](struct_value& fields) {
+          if (fields.size() != element.children.size()) {
+              throw std::runtime_error(fmt::format(
+                "schema/struct mismatch, schema had {} children, struct had {} "
+                "fields. At column {}",
+                element.children.size(),
+                fields.size(),
+                element.column_index));
+          }
+          traversal_levels child_levels = levels;
+          // Increment the definition level as this node in the hierarchy is
+          // defined.
+          ++child_levels.definition_level;
+          for (size_t i = 0; i < element.children.size(); ++i) {
+              struct_field& member = fields[i];
+              const indexed_schema_element& child = element.children[i];
+              cb(child, child_levels, std::move(member.field));
+          }
+      },
+      [&element, levels, cb](null_value& v) {
+          // If the value is `null`, we use the parent definition_level so that
+          // assembly can determine where the `null` started.
+          for (const auto& child : element.children) {
+              cb(child, levels, value(v));
+          }
+      },
+      [&element](list_value&) {
+          throw std::runtime_error(fmt::format(
+            "unexpected list value for non-repeated schema element {}",
+            element.name));
+      },
+      [&element](map_value&) {
+          throw std::runtime_error(fmt::format(
+            "unexpected map value for non-repeated schema element {}",
+            element.name));
+      },
+      [&element](auto& v) {
+          throw std::runtime_error(fmt::format(
+            "unexpected leaf value for optional schema element {}: {}",
+            element.name,
+            value(std::move(v))));
+      });
+}
+
+void for_each_repeated_child(
+  value val,
+  const indexed_schema_element& element,
+  traversal_levels levels,
+  absl::FunctionRef<
+    void(const indexed_schema_element&, traversal_levels, value)> cb) {
+    ss::visit(
+      std::move(val),
+      [&element, levels, cb](null_value& v) {
+          for_each_optional_child(v, element, levels, cb);
+      },
+      [&element, levels, cb](list_value& list) {
+          // Empty lists are equivalent to a `null` value.
+          if (list.empty()) {
+              for_each_optional_child(null_value(), element, levels, cb);
+              return;
+          }
+          traversal_levels child_levels = levels;
+          // We are marking that there is a higher depth here we
+          // need to record, but the repetition_level is only set
+          // *when* we are repeating (so not for the first element).
+          ++child_levels.repetition_depth;
+          for (auto& member : list) {
+              for_each_optional_child(
```

Comment:
I haven't fully wrapped my head around this -- why can we call this and not for_each_child()? Are lists always comprised of structs or nulls?

Reply:
Yeah parquet schema is based on protobuf v2. You can't nest repeated in protobuf. You have to wrap in a group/message for nesting

Reply:
> I haven't fully wrapped my head around this

also this is the most complex part of the algorithm IMO